sys/src/9/amd64/devvq.c

```diff
@@ -0,0 +1,595 @@
+// descriptors of sufficient length. Portions of code are borrowed from the 9front virtio
+// drivers.
+
+static int
+findvdevs(vqctl **vcs)
+{
+	int cnt = 0;
+	// TODO: this seems to work as if MSI-X is not enabled (device conf space starts at 20).
+	// Find out how to deduce msix_enabled from the device.
+	int msix_enabled = 0;
+	Pcidev *p;
+	// Scan the collected PCI devices info, find possible 9p devices
+	for(p = nil; p = pcimatch(p, PCI_VENDOR_ID_REDHAT_QUMRANET, 0);) {
+		if(vcs != nil) {
+			vcs[cnt] = malloc(sizeof(vqctl));
+			if(vcs[cnt] == nil) {
+				print("\nvq: failed to allocate device control structure for device %d\n", cnt);
+				return cnt;
+			}
+			// Use the legacy interface
+			// Allocate the BAR0 I/O space to the driver
+			vqctl *vc = vcs[cnt];
+			vc->pci = p;
+			vc->port = p->mem[0].bar & ~0x1;
+			print("\nvq: device[%d] port %lx\n", cnt, vc->port);
+			char name[32];
+			char *dmap = mapdev(vc);
+			snprint(name, sizeof(name), "%s-%d", dmap?dmap:"virtio", cnt);
+			if(ioalloc(vc->port, p->mem[0].size, 0, name) < 0) {
+				print("\nvq: port %lx in use\n", vc->port);
+				free(vc);
+				vcs[cnt] = nil;
+				return cnt;
+			}
+			// Device reset
+			outb(vc->port + VIRTIO_PCI_STATUS, 0);
+			vc->feat = inl(vc->port + VIRTIO_PCI_HOST_FEATURES);
+			print("\nvq: features %08x\n");
+			outb(vc->port + VIRTIO_PCI_STATUS, VIRTIO_CONFIG_S_ACKNOWLEDGE|VIRTIO_CONFIG_S_DRIVER);
+			int nqs = findvqs(vc->port, 0, nil);
+			print("\nvq: found %d queues\n", nqs);
+			// For each vq allocate and populate its descriptor
+			if(nqs > 0) {
+				vc->vqs = malloc(nqs * sizeof(Virtq *));
+				vc->nqs = nqs;
+				findvqs(vc->port, nqs, vc->vqs);
+				for(int i = 0; i < nqs; i++) {
+					Virtq *q = vc->vqs[i];
+					q->idx = i;
+					q->pdev = vc;
+					coherence();
+					outl(vc->port + VIRTIO_PCI_QUEUE_PFN, PADDR(q->desc)/BY2PG);
+				}
+			}
+			// Device config space contains data in consecutive 8bit input ports
+			vc->dcfgoff = VIRTIO_PCI_CONFIG_OFF(msix_enabled);
+			vc->dcfglen = vc->pci->mem[0].size - vc->dcfgoff;
+			// No process has claimed the device yet
+			vc->ownpid = -1;
+			// Assume that the device config was modified just now
+			vc->dcmtime = -1;
+			// Reset all bits in the IO map
+			vc->vqiomap = 0;
+			// Enable interrupts, complete initialization
+			intrenable(vc->pci->intl, vqintr, vc, vc->pci->tbdf, name);
+			outb(vc->port + VIRTIO_PCI_STATUS, inb(vc->port + VIRTIO_PCI_STATUS) | VIRTIO_CONFIG_S_DRIVER_OK);
+		}
+		cnt++;
+	}
+	return cnt;
+}
+
+// Driver initialization: find all virtio9p devices (by vendor & device ID).
+// Sense the device configuration, figure out mount tags. We assume that the
+// virtio9p devices do not appear or disappear between reboots.
+
+void
+vqinit(void)
+{
+	print("\nvq initializing\n");
+	nvq = findvdevs(nil);
+	print("\nvq: found %d devices\n", nvq);
+	if(nvq == 0) {
+		return;
+	}
+	cvq = malloc(nvq * sizeof(vqctl *));
```

Comment:
Just checking: does this need to be zero'd? if so use mallocz. If there's even a question, use mallocz.